RA_Integration/IRA_Dialog.cpp

```diff
@@ -0,0 +1,320 @@
+#include <Windows.h>
+#include <WindowsX.h>
+#include <CommCtrl.h>
+
+#include "IRA_Dialog.h"
+#include "RA_Core.h"
+
+long IRA_Dialog::storage{ 0 };
+
+// Note to self: Never call DefWindowProc or DefDlgProc for dialogs
+
+IRA_Dialog::IRA_Dialog(int resId, bool isModal) :
+	ResourceId{ resId },
+	is_modal{ isModal }
+{
+	storage = reinterpret_cast<long>(this);
+}
+
+IRA_Dialog::~IRA_Dialog()
+{
+	storage = 0;
+}
+
+INT_PTR IRA_Dialog::DoModal()
+{
+	return DialogBox(g_hThisDLLInst, MAKEINTRESOURCE(ResourceId), g_RAMainWnd, MsgQueue);
+}
+
+HWND IRA_Dialog::Create()
+{
+	return CreateDialog(g_hThisDLLInst, MAKEINTRESOURCE(ResourceId), g_RAMainWnd, MsgQueue);
+}
+
+INT_PTR IRA_Dialog::MsgQueue(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+	IRA_Dialog* ira = reinterpret_cast<IRA_Dialog*>(storage);
+
+	switch ( uMsg )
+	{
+		HANDLE_MSG(hwnd, WM_ACTIVATE, ira->OnActivate);
```

Comment:
shouldn't this just be:
```HANDLE_MSG(hwnd, WM_ACTIVATE, OnActivate);```

Reply:
Because Win32 isn't in C++ it doesn't have the concept of "this" pointer. For "CreateDialog" and "DialogBox" to work, the DialogProc needs to be hooked which means it's gotta be static.

The reason why we need a a pointer is that we need a "virtual pointer" (polymorphism), meaning that any instance derived from IRA_Dialog makes a callback to the message queue.

Reply:
good enough. didn't realize the method was static.